### PR TITLE
Workbench explicit overplot

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work**

Feedback from the January 2018 user meeting gave universal disapproval for the active/hold facility of plot management. Users wanted to decide at the point of plotting the workspace whether it should open in a new figure or overplot on the existing one.

This work removes the active/hold functionality and adds separate items to overplot on to existing figures if they exist.

**To test:**

Start the workbench and either load some data or create some fake data with `CreateSampleWorkspace`. Examine the plot menu of the workspace and check you are happy with the layout and text.

Try all of the options. The behaviour is intended to be as follows:
* plot will always open a new figure
* overplot will add to an existing active figure if one exists and create a new one if one doesn't exist
  * if two or more plot windows are open then clicking on the window makes it active and the next overplot request will go there.

*No associated issue*

**Release Notes** 

*Does not need to be in the release notes. The workbench will be advertised separately.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/applications/workbench/workbench/plotting/globalfiguremanager.py [hunk 1]
```diff
@@ -63,8 +70,14 @@ def destroy(cls, num):
         manager = cls.figs[num]
         manager.canvas.mpl_disconnect(manager._cidgcf)
 
-        if cls._active.num == num:
-            cls._active = None
+        # There must be a good reason for the following careful
+        # rebuilding of the activeQue; what is it?
```
Comment (ianbush) [inline qt/applications/workbench/workbench/plotting/globalfiguremanager.py]: I think this comment is from `_pylab_helpers`, but if I understand this correctly it means we can do a sort by last active in the plot selector.

**Review by ianbush — CHANGES_REQUESTED** [review]
This looks good, I do prefer this over the previous active/hold buttons.

One issue - you can overplot on a colorfill plot. This gives an interesting but not very useful plot :smile:. If possible it would be nice to disable the menu options if the last plot was a colorfill.

At some point it would be good to get this code under test, but that is for another PR.

File: qt/applications/workbench/workbench/plotting/functions.py [hunk 2]
```diff
@@ -86,17 +88,25 @@ def plot(workspaces, spectrum_nums=None, wksp_indices=None, errors=False):
...
     plot_fn = ax.errorbar if errors else ax.plot
     for ws in workspaces:
         for num in nums:
             plot_fn(ws, **{kw: num})
 
     ax.legend()
-    ax.set_title(workspaces[0].name())
+    title = workspaces[0].name()
+    ax.set_title(title)
+    fig.canvas.set_window_title(title)
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/applications/workbench/workbench/plotting/functions.py]: Good point. Matplotlib stores the figure number so we could just use that and append it? The manager requires it to be unique I think...

File: qt/applications/workbench/workbench/plotting/globalfiguremanager.py (hunk 1 above)
Comment (martyngigg) [reply qt/applications/workbench/workbench/plotting/globalfiguremanager.py]: I think you are correct and this will probably be welcome.

**Comment by martyngigg** [comment]
I've added a note to the project board to add tests for the global figure manager and plotting functions. Thanks for reminding me as it got lost in the dash to prepare for the user meeting.

**Comment by martyngigg** [comment]
Thanks for picking up the colorfill oddity. It turns out that MantidPlot did this by accident and I replicated it but we agree that it's not useful and if you drag on to an existing 2D plot it should replace it. Would you agree?

**Comment by martyngigg** [comment]
@ianbush The current code structure of the workspace widget is not playing nicely with this. Our valiant attempts at sharing code mean that the menu generation is really quite disconnected from the rest of the python code controlling the plotting. I have opened #22504 to address this. How would you feel about a simple pop dialog that says you can't overplot a 1D plot on a colorfill plot and asks you to select a new one for now?

File: qt/applications/workbench/workbench/plugins/workspacewidget.py [hunk 3]
```diff
@@ -106,8 +107,14 @@ def _do_plot_spectrum(self, names, errors, overplot):
         :param names: A list of workspace names
         :param errors: If true then error bars will be plotted on the points
         :param overplot: If true then the add to the current figure if one
-                         exists
+                         exists and it is a compatible figure
         """
+        if overplot:
```
Comment (ianbush) [inline qt/applications/workbench/workbench/plugins/workspacewidget.py]: :+1:

**Review by ianbush — APPROVED** [review]
Changes look good now. Plot naming can be improved in #22516 and menu behaviour after #22504.

**Pull request merged**